File: js/src/decode.js
/**
 * TYTX Base - Decoding functions
 *
 * Decode TYTX JSON strings to JavaScript objects.
 *
 * @module decode
 * @license Apache-2.0
 * @copyright Softwell S.r.l. 2025
 */

const { registry } = require('./registry');
require('./types'); // Ensure types are registered

const TYTX_MARKER = '::JS';
const TYTX_PREFIX = 'TYTX://';

// Known type suffixes for scalar detection
const KNOWN_SUFFIXES = new Set(['N', 'D', 'DH', 'DHZ', 'H', 'B', 'L', 'R', 'T']);

/**
 * Hydrate a single typed string value.
 * @param {string} value - String like "100.50::N"
 * @returns {*} Parsed value or original string
 */
function hydrateValue(value) {
    if (typeof value !== 'string' || !value.includes('::')) {
        return value;
    }

    const idx = value.lastIndexOf('::');
    const rawValue = value.slice(0, idx);
    const suffix = value.slice(idx + 2);

    const typeDef = registry.get(suffix);
    if (typeDef) {
        return typeDef.parse(rawValue);
    }

    // Unknown suffix, return as-is
    return value;
}

/**
 * Recursively hydrate typed values in a structure.
 * @param {*} obj
 * @returns {*}
 */
function hydrateRecursive(obj) {
    if (typeof obj === 'string') {
        if (obj.includes('::')) {
            return hydrateValue(obj);
        }
        return obj;
    }

    if (Array.isArray(obj)) {
        return obj.map(item => hydrateRecursive(item));
    }

    if (obj !== null && typeof obj === 'object') {
        const result = {};
        for (const [k, v] of Object.entries(obj)) {
            result[k] = hydrateRecursive(v);
        }
        return result;
    }

    return obj;
}

/**
 * Check if string ends with a valid type suffix.
 * Handles both raw JSON (ends with ::JS) and quoted scalars (ends with ::X").
 * @param {string} data - JSON string
 * @returns {boolean}
 */
function hasTypeSuffix(data) {
    // Handle quoted JSON scalar: "value::X"
    if (data.endsWith('"')) {
        const idx = data.lastIndexOf('::');
        if (idx === -1) return false;
        const suffix = data.slice(idx + 2, -1); // Strip trailing quote
        return KNOWN_SUFFIXES.has(suffix);
    }
    // Handle struct marker: {...}::JS or [...]::JS
    const idx = data.lastIndexOf('::');
    if (idx === -1) return false;
    const suffix = data.slice(idx + 2);
    return KNOWN_SUFFIXES.has(suffix) || suffix === 'JS';
}

/**
 * Decode a TYTX text format string.
 * Handles ::JS suffix and quoted scalars with type suffix.
 *
 * @param {string} data - JSON string with optional ::JS suffix or scalar with type suffix
 * @returns {*} JavaScript object with typed values hydrated
 *
 * @example
 * fromText('{"price":"100.50::N"}::JS')
 * // {price: Big("100.50")}
 *
 * fromText('"2025-01-15::D"')
 * // Date(2025, 0, 15)
 */
function fromText(data) {
    // Strip whitespace (handles trailing newlines, etc.)
    data = data.trim();

    // Check if data has any type suffix
    if (!hasTypeSuffix(data)) {
        // Plain JSON, no TYTX
        return JSON.parse(data);
    }

    // Check for ::JS marker (struct)
    if (data.endsWith(TYTX_MARKER)) {
        data = data.slice(0, -TYTX_MARKER.length);
        const parsed = JSON.parse(data);
        return hydrateRecursive(parsed);
    }

    // Scalar with type suffix (e.g., "2025-01-15::D")
    const parsed = JSON.parse(data);
    // parsed is now a string like "2025-01-15::D", hydrate it
    return hydrateRecursive(parsed);
}

/**
 * Decode a TYTX JSON format string.
 * Handles TYTX:// prefix, ::JS suffix, and quoted scalars.
 *
 * @param {string} data - JSON string with optional TYTX:// prefix
 * @returns {*} JavaScript object with typed values hydrated
 *
 * @example
 * fromJson('TYTX://{"price":"100.50::N"}::JS')
 * // {price: Big("100.50")}
 *
 * fromJson('TYTX://"2025-01-15::D"')
 * // Date(2025, 0, 15)
 */
function fromJson(data) {
    // Strip TYTX:// prefix if present
    if (data.startsWith(TYTX_PREFIX)) {
        data = data.slice(TYTX_PREFIX.length);
    }

    // Delegate to fromText
    return fromText(data);
}

module.exports = {
    fromText,
    fromJson,
    hydrateValue,
    hydrateRecursive,
    TYTX_MARKER,
    TYTX_PREFIX
};

File: src/genro_tytx/decode.py
# Copyright 2025 Softwell S.r.l. - Licensed under Apache License 2.0
"""
TYTX Decoding - TYTX JSON string to Python objects.

Uses json.loads() followed by recursive hydration.
"""

from __future__ import annotations

import json
from typing import Any

from .registry import SUFFIX_TO_TYPE

# Check for orjson availability
try:
    import orjson

    HAS_ORJSON = True
except ImportError:
    HAS_ORJSON = False

TYTX_MARKER = "::JS"
TYTX_PREFIX = "TYTX://"


def _hydrate_value(value: str) -> Any:
    """
    Hydrate a single TYTX-encoded string value (internal).

    Args:
        value: String like "100.50::N" or "2025-01-15::D"

    Returns:
        Python object (Decimal, date, etc.) or original string if not typed
    """
    if "::" not in value:
        return value

    # Find last :: to get suffix
    idx = value.rfind("::")
    raw_value = value[:idx]
    suffix = value[idx + 2 :]

    entry = SUFFIX_TO_TYPE.get(suffix)
    if entry is None:
        # Unknown suffix, return as-is
        return value

    _, deserializer = entry
    return deserializer(raw_value)


def _hydrate_recursive(value: Any) -> Any:
    """Recursively hydrate all typed values in a structure."""
    if isinstance(value, str):
        if "::" in value:
            return _hydrate_value(value)
        return value

    if isinstance(value, dict):
        return {k: _hydrate_recursive(v) for k, v in value.items()}

    if isinstance(value, list):
        return [_hydrate_recursive(item) for item in value]

    return value


def _has_type_suffix(data: str) -> bool:
    """Check if string ends with a valid type suffix (::XXX).

    Handles both raw JSON (ends with ::JS) and quoted scalars (ends with ::X").
    """
    # Handle quoted JSON scalar: "value::X"
    if data.endswith('"'):
        idx = data.rfind("::")
        if idx == -1:
            return False
        suffix = data[idx + 2 : -1]  # Strip trailing quote
        return suffix in SUFFIX_TO_TYPE
    # Handle struct marker: {...}::JS or [...]::JS
    idx = data.rfind("::")
    if idx == -1:
        return False
    suffix = data[idx + 2 :]
    return suffix in SUFFIX_TO_TYPE or suffix == "JS"


def from_text(data: str, *, use_orjson: bool | None = None) -> Any:
    """
    Decode a TYTX JSON string to Python objects.

    Args:
        data: JSON string with ::JS suffix (struct) or ::T suffix (scalar)
        use_orjson: Force orjson (True), stdlib json (False), or auto (None)

    Returns:
        Python object with typed values hydrated

    Example:
        >>> from_text('{"price": "100.50::N"}::JS')
        {"price": Decimal("100.50")}
        >>> from_text('"2025-01-15::D"')
        date(2025, 1, 15)
    """
    # Strip whitespace (handles trailing newlines, etc.)
    data = data.strip()

    if use_orjson is None:
        use_orjson = HAS_ORJSON

    # Check if data has any type suffix
    if not _has_type_suffix(data):
        # Plain JSON, no TYTX
        if use_orjson and HAS_ORJSON:
            return orjson.loads(data)
        return json.loads(data)

    # Check for ::JS marker (struct)
    if data.endswith(TYTX_MARKER):
        data = data[: -len(TYTX_MARKER)]
        if use_orjson and HAS_ORJSON:
            parsed = orjson.loads(data)
        else:
            parsed = json.loads(data)
        return _hydrate_recursive(parsed)

    # Scalar with type suffix (e.g., "2025-01-15::D")
    if use_orjson and HAS_ORJSON:
        parsed = orjson.loads(data)
    else:
        parsed = json.loads(data)

    # parsed is now a string like "2025-01-15::D", hydrate it
    return _hydrate_recursive(parsed)


def from_json(data: str, *, use_orjson: bool | None = None) -> Any:
    """
    Decode a TYTX JSON string with protocol prefix to Python objects.

    Expects TYTX:// prefix per protocol spec.

    Args:
        data: JSON string with TYTX:// prefix and ::JS or ::T suffix
        use_orjson: Force orjson (True), stdlib json (False), or auto (None)

    Returns:
        Python object with typed values hydrated

    Example:
        >>> from_json('TYTX://{"price": "100.50::N"}::JS')
        {"price": Decimal("100.50")}
        >>> from_json('TYTX://"2025-01-15::D"')
        date(2025, 1, 15)
    """
    # Strip whitespace first
    data = data.strip()

    # Strip TYTX:// prefix if present
    if data.startswith(TYTX_PREFIX):
        data = data[len(TYTX_PREFIX) :]

    # Delegate to from_text
    return from_text(data, use_orjson=use_orjson)
